### PR TITLE
Add SAML flow support ADFS endpoint "/adfs/services/trust/13/usernamemixed"

### PR DESCRIPTION
**Pull request by Didainius**
This PR adds support for SAML login using Microsoft Active Directory Services as Identity Provider (IdP). 

It introduces two new `provider` section variables `use_saml_adfs` and `saml_rpt_id` to alter login flow.

## Testing
PR adds a few new variables to configuration struct:
```
SamlUser        string `json:"samlUser,omitempty"`
SamlPassword    string `json:"samlPassword,omitempty"`
SamlCustomRptId string `json:"samlCustomRptId,omitempty"`
```
It allows to triggers explicitly a test `TestAccVcdSamlAuth` designed to try SAML authentication and can be useful if local users are used for main testing.

More details about flow and how it works is in go-vcloud-director SDK PR (https://github.com/vmware/go-vcloud-director/pull/304) as this is where it all is handled. The main point of SDK PR is to make it as easy as possible to use new features.

This PR demonstrates it by showing how little SDK user needs to do.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,8 @@ IMPROVEMENTS:
 * Removed code that handled specific cases for API 29.0 and 30.0. This library now supports VCD versions from 9.5 to 10.1 included [GH-499]
 * Added command line flags to test suite, corresponding to environment variables listed in TESTING.md [GH-505]
 * `resource/vcd_vapp_vm` allows creating VM from multi VM vApp template [GH-501]
+* Add support for SAML auth with Active Directory Federation Services (ADFS) as IdP using
+  "/adfs/services/trust/13/usernamemixed" endpoint. [GH-502]
```
Comment (vbauzysvmware) [inline CHANGELOG.md:10]: [GH-502] -> [GH-504]

File: vcd/auth_saml_test.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+import (
+	"os"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+)
+
+// TestAccVcdSamlAuth explicitly tests a simple operation using SAML auth when explicit SAML
```
Comment (dataclouder) [inline vcd/auth_saml_test.go:13]:
```suggestion
// TestAccVcdSamlAuth tests a simple operation using SAML auth when explicit SAML
```

File: vcd/auth_saml_test.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	backupSysOrg := testConfig.Provider.SysOrg
+	backupEnableConnectionCache := enableConnectionCache
+	enableConnectionCache = false
+	cachedVCDClients.reset()
+
+	t.Logf("Clearing connection cache and switching credentials to: user=%s, use_saml_adfs=%t, rpt_id=%s, org=%s\n",
+		testConfig.Provider.SamlUser, true, testConfig.Provider.SamlCustomRptId, testConfig.VCD.Org)
+
+	// Temporarily override user and password variables
+	_ = os.Setenv("VCD_USER", testConfig.Provider.SamlUser)
```
Comment (dataclouder) [inline vcd/auth_saml_test.go:38]: What happens if `VCD_TOKEN` or `testConfig.Provider.Token` is set? Wouldn't that override the SAML authentication?

File: vcd/config_test.go [hunk 4]
```diff
@@ -34,9 +35,20 @@ import (
...
-		User                     string `json:"user"`
-		Password                 string `json:"password"`
-		Token                    string `json:"token,omitempty"`
+		User            string `json:"user"`
+		Password        string `json:"password"`
+		Token           string `json:"token,omitempty"`
+		UseSamlAdfs     bool   `json:"useSamlAdfs"`
+		CustomAdfsRptId string `json:"customAdfsRptId,omitempty"`
+
+		// The below `SamlUser`, `SamlPassword` and `SamlCustomRptId` variables are optional and are
```
Comment (dataclouder) [inline vcd/config_test.go:44]:
```suggestion
		// The variables `SamlUser`, `SamlPassword` and `SamlCustomRptId` are optional and are
```

File: vcd/sample_vcd_test_config.json [hunk 5]
```diff
@@ -6,6 +6,14 @@
     "user": "root",
     "password": "somePassword",
     "token": "Access token to be used instead of username/password",
+    
+    "//": "Below 3 fields allow to set SAML credentials for tests that specifically use it.",
```
Comment (dataclouder) [inline vcd/sample_vcd_test_config.json:10]:
```suggestion
    "//": "The 3 fields below allow to set SAML credentials for tests that specifically use it.",
```

File: vcd/sample_vcd_test_config.json [hunk 6]
```diff
@@ -6,6 +6,14 @@
...
+    "//": "May be useful when local user credentials are used by default.",
+    "//": "The below credentials will authenticate to Org specified in vcd.org parameter.",
```
Comment (dataclouder) [inline vcd/sample_vcd_test_config.json:12]:
```suggestion
    "//": "These credentials will authenticate to the Org specified in vcd.org parameter.",
```

File: website/docs/index.html.markdown [hunk 7]
```diff
@@ -186,7 +209,17 @@ The following arguments are used to configure the VMware vCloud Director Provide
    instead of username and password. When this is set, username and password will
     be ignored, but should be left in configuration either empty or with any custom values.
     A token can be specified with the `VCD_TOKEN` environment variable.
-    
+
+* `use_saml_adfs` - (Optional) Set `true` to use SAML login flow with Active Directory Federation
```
Comment (lvirbalas) [inline website/docs/index.html.markdown:213]: In discussions about this field. Options so far:
`saml_adfs_enabled`
`saml_adfs_auth_enabled`
`auth_type = "saml_adfs" | "native" (default) | "token" | "ldap"` 

CC: @dataclouder @vbauzysvmware

Comment (vbauzysvmware) [reply website/docs/index.html.markdown:213]: my vote auth_type if possible

File: vcd/config_test.go [hunk 8]
```diff
@@ -34,9 +35,20 @@ import (
...
 type TestConfig struct {
 	Provider struct {
-		User                     string `json:"user"`
-		Password                 string `json:"password"`
-		Token                    string `json:"token,omitempty"`
+		User            string `json:"user"`
+		Password        string `json:"password"`
+		Token           string `json:"token,omitempty"`
+		UseSamlAdfs     bool   `json:"useSamlAdfs"`
+		CustomAdfsRptId string `json:"customAdfsRptId,omitempty"`
```
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/config_test.go:48]: Added

File: vcd/sample_vcd_test_config.json (hunk 6 above)
Comment (Didainius) [reply vcd/sample_vcd_test_config.json]: Changed

File: vcd/sample_vcd_test_config.json (hunk 5 above)
Comment (Didainius) [reply vcd/sample_vcd_test_config.json]: Changed.

File: vcd/config_test.go (hunk 4 above)
Comment (Didainius) [reply vcd/config_test.go]: Changed.

File: vcd/auth_saml_test.go (hunk 2 above)
Comment (Didainius) [reply vcd/auth_saml_test.go]: Changed.

File: CHANGELOG.md (hunk 1 above)
Comment (Didainius) [reply CHANGELOG.md]: Changed.

File: vcd/auth_saml_test.go (hunk 3 above)
Comment (Didainius) [reply vcd/auth_saml_test.go:39]: Well caught! I missed this variable. Changed.

**Comment by vbauzysvmware** [comment]
>     2. Would like to see tests passing for @vbauzysvmware on his ADFS before merging


Replicated ADFS and tested on my env.

**Pull request merged**